File: src/common/tavily-search.ts
import { Logger } from '@nestjs/common';

const logger = new Logger('TavilySearch');

export type WebSearchResult = {
  title: string;
  url: string;
  content: string;
};

/**
 * Real-time web search via Tavily — used to ground Groq-generated answers
 * in current information (e.g. "what UK grants are open right now") that
 * a static system prompt can't know. Best-effort only: returns [] on any
 * failure or missing API key, so callers should treat this purely as an
 * enrichment step, never a hard dependency.
 */
export async function searchWeb(
  query: string,
  options: { maxResults?: number; includeDomains?: string[] } = {},
): Promise<WebSearchResult[]> {
  const apiKey = process.env.TAVILY_API_KEY;
  if (!apiKey) return [];

  try {
    const response = await fetch('https://api.tavily.com/search', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        api_key: apiKey,
        query,
        search_depth: 'basic',
        max_results: options.maxResults ?? 4,
        include_answer: false,
        include_domains: options.includeDomains,
      }),
    });

    if (!response.ok) {
      logger.warn(`Tavily search failed with status ${response.status}`);
      return [];
    }

    const payload = await response.json();
    const results = payload?.results;
    if (!Array.isArray(results)) return [];

    return results
      .filter((r) => r?.title && r?.url)
      .map((r) => ({
        title: String(r.title),
        url: String(r.url),
        content: String(r.content ?? '').slice(0, 600),
      }));
  } catch (error) {
    logger.warn(`Tavily search errored: ${error instanceof Error ? error.message : error}`);
    return [];
  }
}

/** Formats results as a block to drop into a Groq prompt as grounding context. */
export function formatWebResultsForPrompt(results: WebSearchResult[]): string {
  if (results.length === 0) return '';
  return results
    .map((r, i) => `[${i + 1}] ${r.title}\n${r.content}\nSource: ${r.url}`)
    .join('\n\n');
}

File: src/idea-engine/app/services/tavily_search.py
import logging

import httpx

from app.core.config import settings

logger = logging.getLogger(__name__)

TAVILY_URL = "https://api.tavily.com/search"


async def search_web(query: str, max_results: int = 4) -> list[dict]:
    api_key = settings.TAVILY_API_KEY
    if not api_key:
        return []

    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.post(
                TAVILY_URL,
                json={
                    "api_key": api_key,
                    "query": query,
                    "search_depth": "basic",
                    "max_results": max_results,
                    "include_answer": False,
                },
            )

        if response.status_code != 200:
            logger.warning("Tavily search failed with status %s", response.status_code)
            return []

        payload = response.json()
        results = payload.get("results")
        if not isinstance(results, list):
            return []

        return [
            {
                "title": str(r.get("title", "")),
                "url": str(r.get("url", "")),
                "content": str(r.get("content", ""))[:600],
            }
            for r in results
            if r.get("title") and r.get("url")
        ]
    except Exception as e:  # noqa: BLE001 — best-effort, never block generation
        logger.warning("Tavily search errored: %s", e)
        return []


def format_web_results_for_prompt(results: list[dict]) -> str:
    """Formats results as a block to drop into the Groq system prompt as grounding context."""
    if not results:
        return ""

    return "\n\n".join(
        f"[{i + 1}] {r['title']}\n{r['content']}\nSource: {r['url']}"
        for i, r in enumerate(results)
    )
